specification.markdown

```diff
@@ -228,6 +210,95 @@ VINT\_WIDTH | VINT\_MARKER | VINT\_DATA     | Element Data Size Status
         For example -2 can be coded as 0xFFFFFFFFFFFFFE or 0xFFFE or
         0xFE and 5 can be coded 0x000000000005 or 0x0005 or 0x05.
 
+## EBML Element Types
+
+Each defined EBML Element MUST have a declared Element Type. The Element Type defines a concept for storing data that may be constrained by length, endianness, and purpose.
+
+Element Data Type:   Signed Integer
+
+    Endianness:     Big-endian
+    Length:         A Signed Integer Element MUST at least 1 octet in
```

Comment:
I'm not really convinced of that. One if not the most often used way of reading an element goes something like this:

    data_size = read_data_size
    value = 0
    while data_size > 0:
        value += read_byte # (adjust for different data types)
        --data_size

So you always initialize the value anyway, and not only if the data size is != 0.

One thing I sincerely detest about C/C++ is the multitude of cases that cause undefined behavior (like signed integer overflow). Undefined behavior means that the compiler can assume anything, and that leads to so many subtle bugs on occurring in certain cases. I would really like to avoid such cases in EBML if possible. Saying that a data size of 0 is not allowed – what does that mean? That the element itself is invalid and must be ignored? That it can be treated as if it wheren't present? That the parser can return the element containing a random value (e.g. by not initializing its value field unless the data size is > 0)?

I'd like this case to be defined clearly so that such questions cannot even arise. Make it easy for implementers.

Also about the test cases. You can indeed argue that handling `data_size == 0` at the parser level is one more case to test. However, if you forbid this situation the burden of testing moves a level higher. If your parser library doesn't handle such elements then the application above it will probably need some tests for this corner case. If such elements represent undefined behavior then the parser library might simply return an element whose data size is 0 but whose value field is undefined/uninitialized/contains garbarge. An application would have to make sure it can cope with such elements gracefully.

This puts the burden of testing upon many more shoulders (as there's 1:n relationship between the parser library and applications using it) than if it were the parser library's job.

Undefined behavior is bad. In C/C++ there are so many cases of UB only because those languages aim to provide maximum efficiency and have a »don't pay for what you don't use« policy. This is not EBML's concern, though; reliable, extensible and flexible storage is. Those goals conflict with undefined behavior.

Yes, I do feel strongly about that.